### PR TITLE
Feature/602 sri zkp component generate content addressable hash

### PR DESCRIPTION
**Pull request by biscuitdey**
# Description

CRUD management of Content Addressable Hash. The document to be synced is converted into a content addressable hash (CAH) using the crypto library SHA256() function. The CAH is stored in three locations:

- PRIVATE: The document and the CAH are stored on postgres to allow the retrieval of the document using the CAH.
- SHARED: The CAH is also added to the state history merkle tree (Will be added in upcoming PR)
- SHARED: The CAH is also stored on the chosen CCSM.

## Related Issue

#602 : Generate Content Addressable Hash

[Related document](https://docs.google.com/document/d/1mkO5oou0TWk6kVHbTB01dyj7wuNGTfR9w6PAxSgXdGI/edit?usp=sharing)

## Motivation and Context

Generates the Content Addressable Hash of the document that is to be synced amongst the BPI participants.

## How Has This Been Tested

- Unit tests have been written for the API

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:82]: should it not be `CCSMAnchorHash.delete` ? @biscuitdey

**Review by Therecanbeonlyone1969 — COMMENTED** [review]
LGTM apart from the one typo I commented on that should be fixed before merging

**Comment by Therecanbeonlyone1969** [comment]
@biscuitdey ... quick question: I just realized that you are generating the content addressable hash from the document. That is not correct. It must be for the BPIAccount State which contains the proof that the counterparties agreed on a document. The document should be linked to or be part of the BPIAccount state, it is **not** the BPIAccount state since there is no proof of agreement aka synchronization. The content addressable hash should point to the BPIAccount state at the moment in time when that BPIAccount state was created. So updated BPIAccount states have different content addressable hashes. And the BPIAccount state history is organized into a Merkle tree of ordered content addressable hashes. cc @Kasshern @ognjenkurtic @skosito @Manik-Jain @Ybittan

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (skosito) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:82]: yea it looks like this model name is a bit specific and it defaults to camel case making first letter small
i think it should be either `CCSMAnchorHash` or `ccsmAnchorHash`, whatever is easier to configure with prisma

File: examples/bri-3/src/bri/zeroKnowledgeProof/api/err.messages.ts
Comment (skosito) [inline examples/bri-3/src/bri/zeroKnowledgeProof/api/err.messages.ts:5]: not important for now probably but should be split into 2 error messages and 2 error cases maybe for easier debugging?

**Comment by skosito** [comment]
> @biscuitdey ... quick question: I just realized that you are generating the content addressable hash from the document. That is not correct. It must be for the BPIAccount State which contains the proof that the counterparties agreed on a document. The document should be linked to or be part of the BPIAccount state, it is **not** the BPIAccount state since there is no proof of agreement aka synchronization. The content addressable hash should point to the BPIAccount state at the moment in time when that BPIAccount state was created. So updated BPIAccount states have different content addressable hashes. And the BPIAccount state history is organized into a Merkle tree of ordered content addressable hashes. cc @Kasshern @ognjenkurtic @skosito @Manik-Jain @Ybittan

@biscuitdey @Therecanbeonlyone1969 it seems there is a miscommunication with proposal document (in PR description) where we agreed on doing this with basically treating documents as the state, or am i wrong? It looks like we should go back and revisit this with this point in mind and introduce concept of bpi account states too

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:82]: `ccsmAnchorHash` looks more readable to me

File: examples/bri-3/prisma/migrations/20230104084453_add_document_table/migration.sql
Comment (ognjenkurtic) [inline examples/bri-3/prisma/migrations/20230104084453_add_document_table/migration.sql:12]: This is just a discussion point: What does this entity represent? If the BPI serves as the state synchronizer under zero knowledge, we would need to avoid storing customer data in plain text. Or allow it only if the workgroup participants decide to do it. But it makes much more sense to me that the data resides in respective systems of record while sync proofs reside in the BPI. Could you please post a link to the document where this entity design was outlined? I might have missed it completely, sorry.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:82]: what about a brave proposal to ditch ccsm from the name completely and just call it anchorHash everywhere? We all know to what it anchors..

File: examples/bri-3/src/bri/zeroKnowledgeProof/capabilities/createCCSMAnchorHash/createCCSMAnchorHashCommand.handler.ts
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/capabilities/createCCSMAnchorHash/createCCSMAnchorHashCommand.handler.ts:21]: rename document => documentContent

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHash.agent.ts
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHash.agent.ts:9]: inputForProofVerificationparam of method throwErrorIfCCSMAnchorHashInputInvalid should be rename to something like documentContent

File: examples/bri-3/src/bri/zeroKnowledgeProof/api/err.messages.ts
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/api/err.messages.ts:2]: Input for anchor hash creation cannot be empty

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:10]: convention is that storageAgent refers to local storage for all other modules. I would continue using the same convention here and would rename this  CCSMAnchorHashLocalStorageAgent to CCSMAnchorHashStorageAgent. The other one, which is specific becuase it stores on chain, should have a specific unconventional name such as CCSMAnchorHashCcsmStorageAgent ( look at my other comment for ditching ccsm everywhere else, then this name makes more sense ;))

File: examples/bri-3/prisma/migrations/20230104084453_add_document_table/migration.sql
Comment (biscuitdey) [reply examples/bri-3/prisma/migrations/20230104084453_add_document_table/migration.sql:12]: This design for the document is outdated. I will change it. As clarified by Andreas in last week's call and his above comment, we will not hash or store the document directly. We will hash and store the `stateObjectStorage` (which includes the document) parameter of `BPIAccount`, i.e., we are hashing and syncing the BPIAccount state.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:10]: I agree with this and would prefer to ditch ccsm. Using ccsmStorageAgent sounds better.

**Review by Kasshern — COMMENTED** [review]
Since there are quite a few outstanding comments I left my thoughts on a couple of issues but nothing so outstanding needed immediate change for my approval.

**Comment by biscuitdey** [comment]
> @biscuitdey ... quick question: I just realized that you are generating the content addressable hash from the document. That is not correct. It must be for the BPIAccount State which contains the proof that the counterparties agreed on a document. The document should be linked to or be part of the BPIAccount state, it is **not** the BPIAccount state since there is no proof of agreement aka synchronization. The content addressable hash should point to the BPIAccount state at the moment in time when that BPIAccount state was created. So updated BPIAccount states have different content addressable hashes. And the BPIAccount state history is organized into a Merkle tree of ordered content addressable hashes. cc @Kasshern @ognjenkurtic @skosito @Manik-Jain @Ybittan

@Therecanbeonlyone1969  Are we hashing the complete `BPI Account` object or only the `stateObjectStorage` parameter of the `BPI Account` object?

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:10]: @Therecanbeonlyone1969 Can we rename `CCSMAnchorHash` to simply `AnchorHash` and name the blockchain storage agent as `ccsmStorageAgent`?

File: examples/bri-3/src/bri/zeroKnowledgeProof/capabilities/createCCSMAnchorHash/createCCSMAnchorHashCommand.handler.ts
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/capabilities/createCCSMAnchorHash/createCCSMAnchorHashCommand.handler.ts:26]: The ownerAccount id is also passed because it is meta data related to the hash (i.e. who has initiated this state change and hash) and is required for creating the anchor hash object.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHash.agent.ts
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHash.agent.ts:44]: I have abstracted it out because hashing is used in multiple places. Also, in future we might want to use different hash types.

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:10]: My proposal would be the following:

anchorHashStogateAgent - storing into local postgres db
anchorHashCcsmStorageAgent - storing on chain

I would treat the ccsm acronym as every other noun when it comes to casing.

Comment (ognjenkurtic) [reply examples/bri-3/src/bri/zeroKnowledgeProof/agents/ccsmAnchorHashLocalStorage.agent.ts:10]: And +1 on @biscuitdey proposal. Let's ditch ccsm in front of anchor hash completely. It just creates visual noise. @Therecanbeonlyone1969

File: examples/bri-3/src/bri/zeroKnowledgeProof/capabilities/createCCSMAnchorHash/createCCSMAnchorHashCommand.handler.ts
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/zeroKnowledgeProof/capabilities/createCCSMAnchorHash/createCCSMAnchorHashCommand.handler.ts:26]: I was refering to line 26 where the newDocument parameter is passed. Instead of passing that, we should pass command.document since the call to create the new document should happen later in the execution of the command.

**Review by ognjenkurtic — DISMISSED** [review]
Approving since the comments from previous reviews have been applied.  The PR is already too complex because of all the history so i suggest we merge it and continue refining in the upcoming milestone that will specifically deal with state handling.

**Comment by ognjenkurtic** [comment]
@biscuitdey please resolve conflicts and i ll reapprove

**Review by ognjenkurtic — DISMISSED** [review]

File: examples/bri-3/package.json
Comment (Kasshern) [inline examples/bri-3/package.json:42]: @biscuitdey This library seems to be deprecated and built into node modules. Its likely unnecessary

File: examples/bri-3/prisma/schema.prisma
Comment (Kasshern) [inline examples/bri-3/prisma/schema.prisma:118]: This is missing the opposite relation on the BpiSubject model in this schema

**Review by Kasshern — CHANGES_REQUESTED** [review]

File: examples/bri-3/src/bri/zeroKnowledgeProof/agents/anchorHash.agent.ts
Comment (Kasshern) [inline examples/bri-3/src/bri/zeroKnowledgeProof/agents/anchorHash.agent.ts:12]: unimportant but we should all do away with this stock comment going forward

File: examples/bri-3/package.json
Comment (biscuitdey) [reply examples/bri-3/package.json]: No it's not being used.

File: examples/bri-3/package.json
Comment (Kasshern) [inline examples/bri-3/package.json:60]: Still have the types here for bcrypt, but lets just take it out in next PR for sake of getting this through.

**Review by Kasshern — APPROVED** [review]

**Review by ognjenkurtic — APPROVED** [review]

**Pull request merged**
